src/spi.rs

```diff
@@ -0,0 +1,309 @@
+//! Serial Peripheral Interface (SPI) module.
+
+use ::nb;
+use ::hal::spi::{FullDuplex, Mode, Phase, Polarity};
+use ::stm32l4x6::{SPI1, SPI2};
+
+use ::time::Hertz;
+use ::rcc::{APB1, APB2, Clocks};
+
+use ::ptr;
+
+///Describes GPIO Pins to be used by SPI.
+pub mod gpio {
+    ///SPI1 Pins
+    ///
+    ///Uses AF5
+    pub mod spi1 {
+        pub mod nss {
+            pub use ::gpio::{PA4, PA15};
+        }
+        pub mod sck {
+            pub use ::gpio::{PA5, PB3};
+        }
+        pub mod miso {
+            pub use ::gpio::{PA6, PB4};
+        }
+        pub mod mosi {
+            pub use ::gpio::{PA7, PB5};
+        }
+    }
+
+    ///SPI2 Pins
+    ///
+    ///Uses AF5
+    pub mod spi2 {
+        pub mod nss {
+            pub use ::gpio::{PB9, PB12};
+        }
+        pub mod sck {
+            pub use ::gpio::{PB10, PB13};
+        }
+        pub mod miso {
+            pub use ::gpio::{PB14, PC2};
+        }
+        pub mod mosi {
+            pub use ::gpio::{PB15, PC3};
+        }
+    }
+
+    ///SPI3 Pins
+    ///
+    ///Uses AF6.
+    ///Can overlap with others.
+    pub mod spi3 {
+        pub mod nss {
+            pub use ::gpio::{PA4, PA15};
+        }
+
+        pub mod sck {
+            pub use ::gpio::{PB3, PC10};
+        }
+
+        pub mod miso {
+            pub use ::gpio::{PB4, PC11};
+        }
+
+        pub mod mosi {
+            pub use ::gpio::{PB15, PC12};
+        }
+    }
+}
+
+///Describes set of SPI pins
+pub trait Pins<SPI, APB> where Self: Sized {
+    ///Whether pins require remapping
+    const REMAP: bool;
```

Comment:
Hmmm now I think about it, I took it as reference from japaric HAL, but I cannot really find a reference to the need for remapping any SPI.

Could be that all 3 just by confiuring GPIO pins